source/translated_table_model.ts

```diff
@@ -0,0 +1,170 @@
+    } else {
+      this.transactionArray.push(operation);
+    }
+  }
+
+  private processSourceOperation = (operation: Operation) => {
+    if(operation instanceof AddRowOperation) {
+      this.processSourceAdd(operation);
+    } else if(operation instanceof MoveRowOperation) {
+      this.processSourceMove(operation);
+    } else if(operation instanceof RemoveRowOperation) {
+      this.processSourceRemove(operation);
+    } else if(operation instanceof UpdateOperation) {
+      this.processSourceUpdate(operation);
+    } else if(operation instanceof Transaction) {
+      this.beginTransaction();
+      operation.operations.forEach(this.processSourceOperation);
+      this.endTransaction();
+    }
+  }
+
+  private processSourceAdd(operation: AddRowOperation) {
+    const rowIndex = operation.index;
+    const row = []      
+    for(let column = 0; column < this.sourceTable.columnCount; column++) {
+      row.push(this.sourceTable.get(rowIndex, column));
+    }
+    this.sourceRowIndices = this.sourceRowIndices.map(sourceIndex =>
+      sourceIndex >= rowIndex ? sourceIndex + 1 : sourceIndex);
+    this.sourceRowIndices.push(operation.index);
+    this.processOperation(new AddRowOperation(this.rowCount - 1));
+  }
+
+  private processSourceMove(operation: MoveRowOperation) {
+    const sourceIndex = operation.source;
+    const destinationIndex = operation.destination;
+    const translatedSourceIndex = this.sourceRowIndices.
+      findIndex(index => index === sourceIndex);
+    const translatedDestinationIndex = this.sourceRowIndices.
+      findIndex(index => index === destinationIndex);
+    this.sourceRowIndices.splice(translatedDestinationIndex, 0,
+      this.sourceRowIndices.splice(translatedSourceIndex, 1)[0]);
+    this.sourceRowIndices = this.sourceRowIndices.map(translatedIndex => {
+      if(translatedIndex >= destinationIndex &&
+          translatedIndex < sourceIndex) {
+        return translatedIndex + 1;
+      } else if(translatedIndex > sourceIndex &&
+        translatedIndex <= destinationIndex) {
+      return translatedIndex - 1;
+      } else if(translatedIndex === sourceIndex) {
+        return destinationIndex;
+      } else {
+        return translatedIndex;
+      }
+    });
+    this.processOperation(new MoveRowOperation(translatedSourceIndex,
+      translatedDestinationIndex));
+  }
+
+  private processSourceRemove(operation: RemoveRowOperation) {
+    const rowIndex = operation.index;
+    const translatedIndex = this.sourceRowIndices.findIndex(index =>
+      index === rowIndex);
+    this.sourceRowIndices.splice(translatedIndex, 1);
+    this.sourceRowIndices = this.sourceRowIndices.map((sourceIndex, index) =>
+      index >= rowIndex ? sourceIndex - 1 : sourceIndex);
+    this.processOperation(new RemoveRowOperation(translatedIndex));
+  }
+
+  private processSourceUpdate(operation: UpdateOperation) {
+    const translatedIndex = this.sourceRowIndices.findIndex(index =>
```

Comment:
You're using findIndex pretty extensively. For smaller tables, it's not a problem, but the virtual table model is intended to deal with *a lot* of data.

I'd recommend setting up the inverse of sourceRowIndices. It isn't exactly trivial but it will make processing operations a lot faster. Instead of doing findIndex with `O(n)`, with n being the number of rows, you can have a nice quick getTranslatedIndex with `O(1)`.